### DISCUSSION TITLE
Prefetch Next Page in infiniteQuery

### DISCUSSION THREAD

**Discussion by naresh1021** [TanStack/query#2957] [Q&A] [answered]
Is there a way to prefetch the next page in an infiniteQuery. There is a utility on queryClient `prefetchInfiniteQuery` but doesn't mention how to use that to fetch the next set of pages. 

I tried using it with `useEffect` but the result was incorrect. `prefetchInfiniteQuery` was only fetching the data that was already fetched by `useInfiniteQuery`.

Code Snippet:
```
const queryClient = useQueryClient();
const queryKey = 'search';
const queryFn = ({pageParam = 0}) => { ... }

const { data } = useInfiniteQuery(queryKey, queryFn, {
   getNextPageParam: (...) => { ... }
}

useEffect(() => {
   queryClient.prefetchInfiniteQuery(queryKey, queryFn);
}, [data?.pages.length]);
```

When I checked the Network tab, after the prefetchInfiniteQuery, the same data is requested again. Say if the useInfiniteQuery fetched 3 pages, then prefetchInfiniteQuery will also fetch the same amount of data. 

I want the prefetchInfiniteQuery to fetch page 4 instead.

**Comment by TkDodo** [comment, marked as answer]
You can just call `fetchNextPage`, which is returned from useInfiniteQuery.

**Reply by naresh1021** [reply]
With my current implementation, it would cause all the pages to be loaded. All the `useEffect` with `data?.pages.length` as dependencies will call `fetchNextPage` until there are no more pages to load. Now all the pages will be fetched but a prefetch of the next page is required.

**Reply by TkDodo** [reply]
Then you likely need to change the condition. Tbh I’m not sure how prefetching on an infinite query should work. You go to a page that has that query, show the first 25 entries but also load the next 25, which will be immediately shown? Then when should the next 25 be loaded?
The difference to a normal prefetch is that infinite query usually renders everything in one screen, and preferching works best when loading things that other screens would show once you go there (e.g. classic pagination)


What i usually do is just: show 25, and when the user has scrolled to the 20th entry, i load the next 25. So I’m just not waiting until the very end to load more, but do it a bit earlier, thus “prefetching” the next batch before the user has scrolled to the very bottom. With that, users mostly won’t see the loading spinner for the load more

**Comment by jschuur** [comment]
I needed to do something similar, except I only wanted to prefetch the second page, because I wanted to avoid a potential cold start coming from my serverless function that sends the paginated data.

My solution was to immediately make a useQuery request for the second page of data in my custom hook and then modify my useInfiniteQuery's query function to first check if it's being called to request the second page and to return the useQuery data for that second page if it's available, otherwise it does a regular new network request.

```js
const useRecentVideos = () => {
  const { data: secondPage } = useQuery(['prefetchSecondPage'], () => fetchRecentVideos({ pageParam: pageSize }), {
    staleTime: Infinity,
  });

  return useInfiniteQuery(
    ['recentVideos'],
    ({ pageParam }) => (pageParam === pageSize && secondPage ? secondPage : fetchRecentVideos({ pageParam })),
    {
      getNextPageParam: (lastPage) => lastPage.pageInfo.nextOffset,
      staleTime: Infinity,
    }
  );
};
```
I thought this was preferable to @TkDodo's suggestion of loading part of the way through, because of how long the cold start can be vs. how fast users might scroll through the first page of data.
